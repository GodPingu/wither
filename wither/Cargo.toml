[package]
authors = ["Anthony Dodd <Dodd.AnthonyJosiah@gmail.com>"]
categories = ["database", "data-structures"]
description = "An ODM for MongoDB built upon the mongo rust driver."
documentation = "https://docs.rs/wither"
homepage = "https://github.com/thedodd/wither"
keywords = ["mongodb", "odm", "orm"]
license = "Apache-2.0"
name = "wither"
readme = "../README.md"
repository = "https://github.com/thedodd/wither"
version = "0.9.0"
edition = "2018"

[dependencies]
bson = "0.14"
chrono = "0.4"
log = "0.4"
mongodb = "0.9.2"
serde = { version="1", features=["derive"] }
wither_derive = { version = "0.9.0", path = "../wither_derive" }

[dev-dependencies]
lazy_static = "1"

[features]
docinclude = [] # Used only for activating `doc(include="...")` on nightly.

[package.metadata.docs.rs]
features = ["docinclude"] # Activate `docinclude` during docs.rs build.
