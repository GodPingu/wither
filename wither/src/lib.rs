#![cfg_attr(feature="docinclude", feature(external_doc))]
#![cfg_attr(feature="docinclude", doc(include="../README.md"))]

// Re-exports //
pub use mongodb;
pub use mongodb::bson;

pub use wither_derive::Model;
#[cfg(any(feature="sync"))]
pub use wither_derive::ModelSync;

// Common //
mod error;
pub use error::{Result, WitherError};
mod common;
pub use common::IndexModel;

// Async //
mod cursor;
pub use cursor::ModelCursor;

mod migration;
pub use migration::{
    IntervalMigration,
    Migration,
};
mod model;
pub use model::Model;

// Sync //
#[cfg(any(feature="sync"))]
mod sync;
#[cfg(any(feature="sync"))]
pub use sync::ModelCursorSync;

#[cfg(any(feature="sync"))]
pub use sync::{
    IntervalMigrationSync,
    MigrationSync,
};
#[cfg(any(feature="sync"))]
pub use sync::ModelSync;

/// All traits needed for basic usage of the wither system.
pub mod prelude {
    pub use crate::migration::{
        Migrating,
        Migration,
    };
    pub use crate::model::Model;
    pub use wither_derive::Model;

    #[cfg(any(feature="sync"))]
    pub use crate::sync::{
        MigratingSync,
        MigrationSync,
    };
    #[cfg(any(feature="sync"))]
    pub use crate::sync::ModelSync;
    #[cfg(any(feature="sync"))]
    pub use wither_derive::ModelSync;
}
