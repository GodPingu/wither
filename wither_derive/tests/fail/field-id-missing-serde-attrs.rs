use bson::oid::ObjectId;
use serde::{Serialize, Deserialize};
use wither::Model;

#[derive(Serialize, Deserialize, Model)]
struct BadModel {
    id: Option<ObjectId>,
}

fn main() {}
