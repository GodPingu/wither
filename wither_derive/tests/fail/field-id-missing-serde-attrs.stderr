error: the ID field of wither::Models must have the attribute `#[serde(rename="_id", skip_serializing_if="Option::is_none")]`
 --> $DIR/field-id-missing-serde-attrs.rs:7:5
  |
7 |     id: Option<ObjectId>,
  |     ^^
